skimage/transform/_distance_transform.pyx

```diff
@@ -0,0 +1,143 @@
+    return <double>np.finfo(np.float64).min
+
+def _generalized_distance_transform_1d_euclidean(double[:] arr, double[:] cost_arr,
+                                       bint isfirst, double[::1] domains,
+                                       Py_ssize_t[::1] centers, double[::1] out):
+    cdef Py_ssize_t length = len(arr)
+    cdef Py_ssize_t i, rightmost, current_domain
+    cdef double intersection
+
+    if isfirst:
+        for i in range(length):
+            cost_arr[i] = f(<bint>arr[i])
+
+    rightmost = 0
+    domains[0] = -np.inf
+    domains[1] = np.inf
+    
+    for i in range(1,length):
+        intersection = euclidean_meet(i,centers[rightmost],cost_arr)
+        while intersection <= domains[rightmost]:
+            rightmost-=1
+            intersection = euclidean_meet(i,centers[rightmost],cost_arr)
+
+        rightmost+=1
+        centers[rightmost]=i
+        domains[rightmost]=intersection
+        domains[rightmost+1] = np.inf
+
+    current_domain = 0
+
+    for i in range(length):
+        while domains[current_domain+1]<i:
+            current_domain += 1
+        out[i] = euclidean_dist(i,centers[current_domain],cost_arr[centers[current_domain]])
+    return out
+
+def _generalized_distance_transform_1d_manhattan(double[:] arr, double[:] cost_arr,
+                                       bint isfirst, double[::1] domains,
+                                       Py_ssize_t[::1] centers, double[::1] out):
+    cdef Py_ssize_t length = len(arr)
+    cdef Py_ssize_t i, rightmost, current_domain
+    cdef double intersection
+
+    if isfirst:
+        for i in range(length):
+            cost_arr[i] = f(<bint>arr[i])
+
+    rightmost = 0
+    domains[0] = -np.inf
+    domains[1] = np.inf
+    
+    for i in range(1,length):
+        intersection = manhattan_meet(i,centers[rightmost],cost_arr)
+        while intersection <= domains[rightmost]:
+            rightmost-=1
+            intersection = manhattan_meet(i,centers[rightmost],cost_arr)
+
+        rightmost+=1
+        centers[rightmost]=i
+        domains[rightmost]=intersection
+        domains[rightmost+1] = np.inf
+
+    current_domain = 0
+
+    for i in range(length):
+        while domains[current_domain+1]<i:
+            current_domain += 1
+        out[i] = manhattan_dist(i,centers[current_domain],cost_arr[centers[current_domain]])
+    return out
+
+"""
+def _generalized_distance_transform_1d_slow(double[:] arr, double[:] cost_arr,
+                                       cost_func, dist_func, dist_meet,
+                                       bint isfirst, double[::1] domains,
+                                       Py_ssize_t[::1] centers, double[::1] out):"""
+def _generalized_distance_transform_1d_slow(arr,cost_arr,
+                                       cost_func, dist_func, dist_meet,
+                                       bint isfirst, domains,
+                                       centers, out):
+    cdef Py_ssize_t length = len(arr)
```

Comment:
you have might a hard time releasing the GIL here because `arr` is a python object. Sometimes to get around this, we define a `c_arr` that is a memory view of the python array. We have a few examples of those. We can get to this one later though because it is obvious it gave you a hard time before.

Reply:
its easier than that. memoryviews have lengths. Are you ready to release the GIL? if so, lets just go over the changes in the real PR.